_posts/2021-12-29-readable-tests-by-example.md

```diff
@@ -0,0 +1,552 @@
+according to the running promotional campaign (current price list configuration), the order will be delivered free
+of charge.
+
+Additionally, for the VIP customer, a free music track should be sent to their mailbox after the payment of the order.
+
+After paying for the order, no modifications can be made.
+```
+
+Such assumptions can be translated into one of the BDD-style scenarios:
+<img alt="API package structure" src="/img/articles/2021-12-29-readable-tests-by-example/1.png" title="Scenario 1.1">
+Scenario 1.1
+
+### Architecture
+
+The structure of the code reflects the architecture that was adopted during the implementation works. The application is
+a modular monolith written based on the architectural style
+of [Ports & Adapter](https://blog.allegro.tech/2020/05/hexagonal-architecture-by-example.html), as well as
+the [Domain Driven Design](https://www.dddcommunity.org/learning-ddd/what_is_ddd/) approach.
+
+<img alt="API package structure" src="/img/articles/2021-12-29-readable-tests-by-example/2.png"/>
+
+We can distinguish the following packages:
+
+* `catalogue`: reflects the catalogue of products with their unit prices
+
+* `client`: provides information on client scoring (VIP, STANDARD)
+
+* `common`: contains common concepts, objects that appear in other contexts, e.g. the Money class
+
+* `delivery`: calculates the delivery price based on the defined policy
+
+* `order`: keeps the logic related to the user’s order, such as the amount of payment, or the ability
+  to find them among other orders
+
+* `sales`: provides information on types promotions (e.g. price list configuration), especially the minimum value order
+  needed for free delivery
+
+As the order-related domain is the most complex, it actually coordinates the entire purchasing process, and thus it
+should provide a comprehensive example for our further consideration.
+
+## How can we test such code?
+
+Each design must feature certain guidelines, which is also the case in choosing the testing methodology. In IT
+literature, we can probably find many interesting references to how we should test an application code, nevertheless, in
+this article I am going to show you how, on a daily basis, me and other teams at Allegro approach this issue in our
+work.
+
+Below I have listed the main assumptions that will guide us throughout the rest of the article:
+
+* Tests should be a living documentation describing the real requirements in the form of clear scenarios and should be
+  easy to understand for every person who joins the project;
+
+* Application will be tested through a black-box approach that examines the functionality of the application without
+  looking into its internal structures or behavior. Thanks to this we focus on the functionality of what we are testing
+  and not on the details of the implementation itself;
+
+* Adding a new test to the existing scenario should not be difficult, and should be based on the existing ready-to-use
```

Comment:
Tutaj mamy powtórzenie słówka `should`. Na chwile obecną nie mam na szybko pomysłu o jego zastąpieniu.
@kzadurska  Jak coś wrzucam taki komentarz, abym nie zapomniał o tym :)

Reply:
Mnie jakoś specjalnie nie razi, ale może ktoś z reviewersów będzie miał jakąś propozycję 😉